### PR TITLE
[CURA-1923] UpdateChecker: Using jobs for checking and cleanups

### PR DESCRIPTION
**Pull request by thopiekar**
Does the same like in SliceInfo in Cura.
Additionally I replaced the current usage of threading with UM.Job.

Contributes to CURA-1923

### CODE REVIEW

File: plugins/UpdateChecker/UpdateChecker.py [hunk 1]
```diff
@@ -9,54 +9,35 @@
...
         Logger.log("i", "Checking for new version of %s" % application_name)
 
         try:
-            latest_version_file = urllib.request.urlopen("http://software.ultimaker.com/latest.json")
-        except Exception as e:
-            Logger.log("e", "Failed to check for new version. %s" %e)
-            if not silent:
+            latest_version_file = urllib.request.urlopen(self.url)
+        except Exception:
+            Logger.logException("e", "Failed to check for new version.")
```
Comment (fieldOfView) [inline plugins/UpdateChecker/UpdateChecker.py]: Any reason in particular why you are no longer showing what the exception is?

**Comment by fieldOfView** [comment]
Though switching from Threading.Thread to Uranium.Jobs has cleaned up the code, unfortunately I still get the delay (and the exception logging has become a lot scarier to look at):
```
2016-07-28 08:03:19,489 - INFO - ...\UM\PluginRegistry.py (loadPlugin [62]): Loaded plugin UpdateChecker
2016-07-28 08:03:19,489 - INFO - ...\UM\..\plugins\UpdateChecker\UpdateChecker.py (run [28]): Checking for new version of cura
2016-07-28 08:03:19,538 - INFO - ...\UM\PluginRegistry.py (loadPlugin [62]): Loaded plugin SimpleView
2016-07-28 08:03:19,559 - INFO - ...\UM\PluginRegistry.py (loadPlugin [62]): Loaded plugin WireframeView
2016-07-28 08:03:19,585 - INFO - ...\UM\PluginRegistry.py (loadPlugin [62]): Loaded plugin 3MFReader
2016-07-28 08:03:19,611 - INFO - ...\UM\PluginRegistry.py (loadPlugin [62]): Loaded plugin AutoSave
2016-07-28 08:03:19,643 - INFO - ...\UM\PluginRegistry.py (loadPlugin [62]): Loaded plugin ChangeLogPlugin
2016-07-28 08:03:19,645 - WARNING - ...\UM\PluginRegistry.py (loadPlugin [62]): Plugin CuraEngineBackend was already loaded
2016-07-28 08:03:19,678 - INFO - ...\UM\PluginRegistry.py (loadPlugin [62]): Loaded plugin CuraProfileReader
2016-07-28 08:03:19,711 - INFO - ...\UM\PluginRegistry.py (loadPlugin [62]): Loaded plugin CuraProfileWriter
2016-07-28 08:03:19,747 - INFO - ...\UM\PluginRegistry.py (loadPlugin [62]): Loaded plugin GCodeProfileReader
2016-07-28 08:03:19,780 - INFO - ...\UM\PluginRegistry.py (loadPlugin [62]): Loaded plugin GCodeWriter
2016-07-28 08:03:19,820 - INFO - ...\UM\PluginRegistry.py (loadPlugin [62]): Loaded plugin ImageReader
2016-07-28 08:03:31,517 - ERROR - ...\UM\Logger.py (logException [49]): Exception: Failed to check for new version.
2016-07-28 08:03:31,615 - INFO - ...\UM\PluginRegistry.py (loadPlugin [62]): Loaded plugin JediWifiPrintingPlugin
2016-07-28 08:03:31,649 - ERROR - ...\UM\Logger.py (logException [49]): Traceback (most recent call last):
2016-07-28 08:03:31,657 - ERROR - ...\UM\Logger.py (logException [49]):   File "C:\Users\Aldo\Python34\lib\urllib\request.py", line 1189, in do_open
2016-07-28 08:03:31,662 - ERROR - ...\UM\Logger.py (logException [49]):     h.request(req.get_method(), req.selector, req.data, headers)
2016-07-28 08:03:31,671 - ERROR - ...\UM\Logger.py (logException [49]):   File "C:\Users\Aldo\Python34\lib\http\client.py", line 1090, in request
2016-07-28 08:03:31,686 - ERROR - ...\UM\Logger.py (logException [49]):     self._send_request(method, url, body, headers)
2016-07-28 08:03:31,696 - ERROR - ...\UM\Logger.py (logException [49]):   File "C:\Users\Aldo\Python34\lib\http\client.py", line 1128, in _send_request
2016-07-28 08:03:31,701 - ERROR - ...\UM\Logger.py (logException [49]):     self.endheaders(body)
2016-07-28 08:03:31,708 - ERROR - ...\UM\Logger.py (logException [49]):   File "C:\Users\Aldo\Python34\lib\http\client.py", line 1086, in endheaders
2016-07-28 08:03:31,713 - ERROR - ...\UM\Logger.py (logException [49]):     self._send_output(message_body)
2016-07-28 08:03:31,718 - ERROR - ...\UM\Logger.py (logException [49]):   File "C:\Users\Aldo\Python34\lib\http\client.py", line 924, in _send_output
2016-07-28 08:03:31,728 - ERROR - ...\UM\Logger.py (logException [49]):     self.send(msg)
2016-07-28 08:03:31,733 - ERROR - ...\UM\Logger.py (logException [49]):   File "C:\Users\Aldo\Python34\lib\http\client.py", line 859, in send
2016-07-28 08:03:31,735 - ERROR - ...\UM\Logger.py (logException [49]):     self.connect()
2016-07-28 08:03:31,737 - ERROR - ...\UM\Logger.py (logException [49]):   File "C:\Users\Aldo\Python34\lib\http\client.py", line 836, in connect
2016-07-28 08:03:31,738 - ERROR - ...\UM\Logger.py (logException [49]):     self.timeout, self.source_address)
2016-07-28 08:03:31,739 - ERROR - ...\UM\Logger.py (logException [49]):   File "C:\Users\Aldo\Python34\lib\socket.py", line 491, in create_connection
2016-07-28 08:03:31,740 - ERROR - ...\UM\Logger.py (logException [49]):     for res in getaddrinfo(host, port, 0, SOCK_STREAM):
2016-07-28 08:03:31,742 - ERROR - ...\UM\Logger.py (logException [49]):   File "C:\Users\Aldo\Python34\lib\socket.py", line 530, in getaddrinfo
2016-07-28 08:03:31,743 - ERROR - ...\UM\Logger.py (logException [49]):     for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
2016-07-28 08:03:31,745 - ERROR - ...\UM\Logger.py (logException [49]): socket.gaierror: [Errno 11004] getaddrinfo failed
2016-07-28 08:03:31,746 - ERROR - ...\UM\Logger.py (logException [49]):
2016-07-28 08:03:31,749 - ERROR - ...\UM\Logger.py (logException [49]): During handling of the above exception, another exception occurred:
2016-07-28 08:03:31,752 - ERROR - ...\UM\Logger.py (logException [49]):
2016-07-28 08:03:31,759 - ERROR - ...\UM\Logger.py (logException [49]): Traceback (most recent call last):
2016-07-28 08:03:31,765 - ERROR - ...\UM\Logger.py (logException [49]):   File "C:\Users\Aldo\Documents\CodePr~1\UM\Uranium\UM\..\plugins\UpdateChecker\UpdateChecker.py", line 37, in run
2016-07-28 08:03:31,770 - ERROR - ...\UM\Logger.py (logException [49]):     latest_version_file = urllib.request.urlopen(self.url)
2016-07-28 08:03:31,778 - ERROR - ...\UM\Logger.py (logException [49]):   File "C:\Users\Aldo\Python34\lib\urllib\request.py", line 153, in urlopen
2016-07-28 08:03:31,785 - ERROR - ...\UM\Logger.py (logException [49]):     return opener.open(url, data, timeout)
2016-07-28 08:03:31,788 - ERROR - ...\UM\Logger.py (logException [49]):   File "C:\Users\Aldo\Python34\lib\urllib\request.py", line 455, in open
2016-07-28 08:03:31,796 - ERROR - ...\UM\Logger.py (logException [49]):     response = self._open(req, data)
2016-07-28 08:03:31,803 - ERROR - ...\UM\Logger.py (logException [49]):   File "C:\Users\Aldo\Python34\lib\urllib\request.py", line 473, in _open
2016-07-28 08:03:31,811 - ERROR - ...\UM\Logger.py (logException [49]):     '_open', req)
2016-07-28 08:03:31,819 - ERROR - ...\UM\Logger.py (logException [49]):   File "C:\Users\Aldo\Python34\lib\urllib\request.py", line 433, in _call_chain
2016-07-28 08:03:31,826 - ERROR - ...\UM\Logger.py (logException [49]):     result = func(*args)
2016-07-28 08:03:31,834 - ERROR - ...\UM\Logger.py (logException [49]):   File "C:\Users\Aldo\Python34\lib\urllib\request.py", line 1215, in http_open
2016-07-28 08:03:31,841 - ERROR - ...\UM\Logger.py (logException [49]):     return self.do_open(http.client.HTTPConnection, req)
2016-07-28 08:03:31,846 - ERROR - ...\UM\Logger.py (logException [49]):   File "C:\Users\Aldo\Python34\lib\urllib\request.py", line 1192, in do_open
2016-07-28 08:03:31,853 - ERROR - ...\UM\Logger.py (logException [49]):     raise URLError(err)
2016-07-28 08:03:31,861 - ERROR - ...\UM\Logger.py (logException [49]): urllib.error.URLError: <urlopen error [Errno 11004] getaddrinfo failed>
```

Note the 12 second hole between loading ImageReader and the exception about the update check.

I think I liked the logging of the exception better before, because now we get 50 lines of red ERROR entries.

File: plugins/UpdateChecker/UpdateChecker.py (hunk 1 above)
Comment (thopiekar) [inline plugins/UpdateChecker/UpdateChecker.py]: Logger will do this for us. When looking at logException() it says that you only need to pass an message and the last message of an exception will be taken from the traceback module.
See: https://github.com/Ultimaker/Uranium/blob/master/UM/Logger.py#L53

So in my point of view no need to take the exception and pass it to the Logger.

Comment (thopiekar) [inline plugins/UpdateChecker/UpdateChecker.py]: Oops, just saw the "Edit:" to late ^^

**Comment by thopiekar** [comment]
But is the UX freezing for these 12s?
The urllib functions will still run and cause timeout errors, but now they shouldn't block. Additionally we could add a "UM.Network" module and add a function there "hasInternetConnection()". There are DBus requests you can send to NetworkManager or ConnMan (same for Windows and OSX) we could check first before trying to send/receive data from the internet.
But at the moment keeping the tasks at the background should be enough.

Hmm, but then it is a logger issue that it returns to much or we'll need to pass something to it to work like expected.

**Comment by fieldOfView** [comment]
Yes, the application takes 10+ seconds longer to load when I have no network connection (in the train to work). I don't know/understand why, because the check is happening in its own thread.

I did test if it was something with the ImageReader plugin somehow waiting for threads, but if I just remove that plugin then the pause will be after the next plugin.

**Comment by fieldOfView** [comment]
I would very much prefer this code not to use logException. The logException output is unnecessarily verbose here. It has its use when we would need to find out what went wrong exactly, but here a simple log error "could not contact the updates server: timeout"  or something like that would be much better. We know what goes wrong here in what order.

Also while the use of Jobs vs Threads cleans up the code and is more along the "Uranium way of doing things", it does not seem to solve the problem for me.

**Comment by thopiekar** [comment]
Due to problems I have here with "Cura" <=> "CuraEngine", I can't test it even.

Are the internal builds working fine? Do you get this https://github.com/Ultimaker/Cura/issues/899 , too?

**Comment by Ghostkeeper** [comment]
Internal builds are working.

**Comment by thopiekar** [comment]
Found the problem. The engine was outdated again..

Made a PR in Cura to use the nightly engine installed from my PPA, so I
don't have to copy it into bin/ everytime.

Mit freundlichen Grüßen/With kind regards
- Thomas Karl Pietrowski

**Comment by Ghostkeeper** [comment]
I've made a symlink from bin to my CuraEngine build to do the "updating" for me :smiley:

**Pull request merged**